src/category_theory/abelian/homology.lean

```diff
@@ -0,0 +1,260 @@
+
+namespace category_theory.abelian
+
+/-- The cokernel of `kernel.lift g f w`. This is isomorphic to `homology f g w`.
+  See `homology_iso_cokernel_lift`. -/
+abbreviation homology_c : A :=
+cokernel (kernel.lift g f w)
+
+/-- The kernel of `cokernel.desc f g w`. This is isomorphic to `homology f g w`.
+  See `homology_iso_kernel_desc`. -/
+abbreviation homology_k : A :=
+kernel (cokernel.desc f g w)
+
+/-- The canonical map from `homology_c` to `homology_k`.
+  This is an isomorphism, and it is used in obtaining the API for `homology f g w`
+  in the bottom of this file. -/
+abbreviation homology_c_to_k : homology_c f g w ⟶ homology_k f g w :=
+cokernel.desc _ (kernel.lift _ (kernel.ι _ ≫ cokernel.π _) (by simp)) begin
+  apply limits.equalizer.hom_ext,
+  simp,
+end
+
+local attribute [instance] pseudoelement.hom_to_fun pseudoelement.has_zero
+
+instance : mono (homology_c_to_k f g w) :=
+begin
+  apply pseudoelement.mono_of_zero_of_map_zero,
+  intros a ha,
+  obtain ⟨a,rfl⟩ := pseudoelement.pseudo_surjective_of_epi (cokernel.π (kernel.lift g f w)) a,
+  apply_fun (kernel.ι (cokernel.desc f g w)) at ha,
+  simp only [←pseudoelement.comp_apply, cokernel.π_desc,
+    kernel.lift_ι, pseudoelement.apply_zero] at ha,
+  simp only [pseudoelement.comp_apply] at ha,
+  haveI : exact f (cokernel.π f) := exact_cokernel f,
+  obtain ⟨b,hb⟩ : ∃ b, f b = _ := pseudoelement.pseudo_exact_of_exact.2 _ ha,
+  suffices : ∃ c, kernel.lift g f w c = a,
+  { obtain ⟨c,rfl⟩ := this,
+    simp [← pseudoelement.comp_apply] },
+  use b,
+  apply_fun kernel.ι g,
+  swap, { apply pseudoelement.pseudo_injective_of_mono },
+  simpa [← pseudoelement.comp_apply]
+end
+
+instance : epi (homology_c_to_k f g w) :=
+begin
+  apply pseudoelement.epi_of_pseudo_surjective,
+  intros a,
+  let b := kernel.ι (cokernel.desc f g w) a,
+  haveI : exact f (cokernel.π f) := exact_cokernel f,
+  obtain ⟨c,hc⟩ : ∃ c, cokernel.π f c = b,
+    apply pseudoelement.pseudo_surjective_of_epi (cokernel.π f),
+  have : g c = 0,
+  { dsimp [b] at hc,
+    rw [(show g = cokernel.π f ≫ cokernel.desc f g w, by simp), pseudoelement.comp_apply, hc],
+    simp [← pseudoelement.comp_apply] },
+  obtain ⟨d,hd⟩ : ∃ d, kernel.ι g d = c,
+  { apply pseudoelement.pseudo_exact_of_exact.2 _ this,
+    apply exact_kernel_ι },
+  use cokernel.π (kernel.lift g f w) d,
+  apply_fun kernel.ι (cokernel.desc f g w),
+  swap, { apply pseudoelement.pseudo_injective_of_mono },
+  simp only [←pseudoelement.comp_apply, cokernel.π_desc, kernel.lift_ι],
+  simp only [pseudoelement.comp_apply, hd, hc],
+end
+
+instance (w : f ≫ g = 0) : is_iso (homology_c_to_k f g w) := is_iso_of_mono_of_epi _
+
+end category_theory.abelian
+
+/-- The homology associated to `f` and `g` is isomorphic to a kernel. -/
+def homology_iso_kernel_desc : homology f g w ≅ kernel (cokernel.desc f g w) :=
+homology_iso_cokernel_lift _ _ _ ≪≫ as_iso (category_theory.abelian.homology_c_to_k _ _ _)
```

Comment:
I think this is an example of what I was referring to.
For example simp lemmas like
```
lemma homology_iso_kernel_desc_hom_ι :
  (homology_iso_kernel_desc _ _ _).hom ≫ kernel.ι _ = homology.ι f g w := rfl

lemma homology_iso_kernel_desc_inv_ι :
  (homology_iso_kernel_desc _ _ _).inv ≫ homology.ι f g w = kernel.ι _ := sorry
``` 
And the analogous lemmas for `homology.π'`. IMO these come up quite often when using
`homology.hom_to_ext` or `kernel.ext` et cetera.

Reply:
I'm not convinced about this. The isomorphism `homology_iso_kernel_desc` should be merely an implementation detail and shouldn't actually come up at all. And in the case where it does come up, it's probably because the user actually wanted to unfold something, so such `simp` lemmas might be counter productive.